File: BonusServer_Node/configSetting.js

import fs from "fs";
import Log from "./utils/log.js";

const mConfigJson = {};
let mConfigFileName = null;

export default class ConfigSetting
{
    static BETWIN_RULE = {
        Rule_0: 0,
        Rule_1: 1
    }

    constructor()
    {
    }

    // 載入配置文件
    static async loadConfig(filename)
    {
        try {
            const data = fs.readFileSync(filename, 'utf8');
            const config = JSON.parse(data);
            
            mConfigJson.Protocol = config.hasOwnProperty('Protocol') ? config.Protocol : 'http';
            mConfigJson.ExportLog = config.ExportLog ? true : false;
            mConfigJson.LanguageType = config.hasOwnProperty('LanguageType') ? parseInt(config.LanguageType) : 0;
            mConfigJson.databases = config.hasOwnProperty('databases') ? config.databases : [];
            mConfigJson.BonusServerDomain = config.hasOwnProperty('BonusServerDomain') ? config.BonusServerDomain : '127.0.0.1';
            mConfigJson.BonusServerPort = config.hasOwnProperty('BonusServerPort') ? parseInt(config.BonusServerPort) : 80;
            mConfigJson.UpperDomain = config.UpperDomain;
            mConfigJson.SubDomains = [];
            if (config.SubDomains) {
                const subDomains = config.SubDomains;
                for (let i = 0; i < subDomains.length; i++)
                {
                    const obj = subDomains[i];
                    if (!obj) continue;

                    const domain = obj.domain;
                    const password = obj.password;

                    mConfigJson.SubDomains.push({
                        Domain: domain ? domain : "",
                        Password: password ? password : "" 
                    });
                }
            }
            mConfigJson.TransferScores = config.hasOwnProperty('TransferScores') ? config.TransferScores : 'http://localhost/APIs/TransferPoints.ashx';
            mConfigJson.CollectSubScale = config.hasOwnProperty('CollectSubScale') ? parseFloat(config.CollectSubScale) : 1.0;
            mConfigJson.RabbitMQServer = config.hasOwnProperty('RabbitMQServer') ? config.RabbitMQServer : '127.0.0.1';
            mConfigJson.RabbitMQUserName = config.hasOwnProperty('RabbitMQUserName') ? config.RabbitMQUserName : 'guest';
            mConfigJson.RabbitMQPassword = config.hasOwnProperty('RabbitMQPassword') ? config.RabbitMQPassword : 'guest';
            mConfigJson.BonusServerPassword = config.hasOwnProperty('BonusServerPassword') ? config.BonusServerPassword : 'localhost';
            if (config.hasOwnProperty('BetWinRule')) {
                const ruleId = parseInt(config.BetWinRule)
                switch (ruleId) {
                    case 0:
                        mConfigJson.BetWinRule = ConfigSetting.BETWIN_RULE.Rule_0;
                        break;
                    case 1:
                    default:
                        mConfigJson.BetWinRule = ConfigSetting.BETWIN_RULE.Rule_1;
                        break;
                }
            }
            else {
                mConfigJson.BetWinRule = ConfigSetting.BETWIN_RULE.Rule_1;
            }
            mConfigJson.WebSite = config.hasOwnProperty('WebSite') ? config.WebSite : '彩金測試場';
            try {
                if (config.hasOwnProperty('ConditionWinA') == false) throw new Error("no settings for ConditionWinA");
                // confirm settings
                const obj = JSON.parse(config.ConditionWinA);
                mConfigJson.ConditionWinA = JSON.stringify(obj);
            }
            catch(err) {
                switch(mConfigJson.BetWinRule) {
                    case ConfigSetting.BETWIN_RULE.Rule_0:                    
                        mConfigJson.ConditionWinA = "{\"ScoreInterval\":3000.0,\"WinOverBet\":0.9,\"WinCount\":300,\"MinPay\":1}";
                        break;
                    case ConfigSetting.BETWIN_RULE.Rule_1:
                    default:
                        mConfigJson.ConditionWinA = "{\"ScoreInterval\":3000.0,\"WinRatio\":0.002,\"MinTotalBet\":0.0,\"MinPay\":1}";
                        break;
                }
            }
            try {
                if (config.hasOwnProperty('ConditionWinB') == false) throw new Error("no settings for ConditionWinB");
                // confirm settings
                const obj = JSON.parse(config.ConditionWinB);
                mConfigJson.ConditionWinB = JSON.stringify(obj);
            }
            catch(err) {
                switch(mConfigJson.BetWinRule) {
                    case ConfigSetting.BETWIN_RULE.Rule_0:                    
                        mConfigJson.ConditionWinB = "{\"ScoreInterval\":3000.0,\"WinOverBet\":0.7,\"WinCount\":100,\"MinPay\":1}";
                        break;
                    case ConfigSetting.BETWIN_RULE.Rule_1:
                    default:
                        mConfigJson.ConditionWinB = "{\"ScoreInterval\":3000.0,\"WinRatio\":0.006,\"MinTotalBet\":0.0,\"MinPay\":1}";
                        break;
                }
            }
            try {
                if (config.hasOwnProperty('ConditionWinCR') == false) throw new Error("no settings for ConditionWinCR");
                // confirm settings
                const obj = JSON.parse(config.ConditionWinCR);
                mConfigJson.ConditionWinCR = JSON.stringify(obj);
            }
            catch(err) {
                switch(mConfigJson.BetWinRule) {
                    case ConfigSetting.BETWIN_RULE.Rule_0:                    
                        mConfigJson.ConditionWinCR = "{\"ScoreInterval\":3000.0,\"WinOverBet\":0.9,\"WinCount\":300,\"MinPay\":1}";
                        break;
                    case ConfigSetting.BETWIN_RULE.Rule_1:
                    default:
                        mConfigJson.ConditionWinCR = "{\"ScoreInterval\":3000.0,\"WinRatio\":0.002,\"MinTotalBet\":0.0,\"MinPay\":1}";
                        break;
                }
            }
            mConfigJson.SSL_PATH = config.hasOwnProperty('SSL_PATH') ? config.SSL_PATH : null;

            // store filename
            mConfigFileName = filename;
            return mConfigJson;
        }
        catch (error) {
            Log.storeMsg(`Error loading config.bin: ${error.message}. Ensure the file exists and has the correct format.`);
        }
    };

    static async reload()
    {
        try {
            if (!mConfigFileName) throw new Error("invalid filename");
            await ConfigSetting.loadConfig(mConfigFileName);
        }
        catch(err) {
            const errMsg = err.message ? err.message : err;
            console.error(`reload() got ${errMsg}`);
        }
    }

    static async storeConfig(objParams)
    {
        try {
            if (!mConfigFileName) throw new Error("invalid filename");

            if (objParams.hasOwnProperty('BonusServerDomain')) mConfigJson.BonusServerDomain = objParams.BonusServerDomain;
            if (objParams.hasOwnProperty('BonusServerPort')) mConfigJson.BonusServerPort = objParams.BonusServerPort;
            if (objParams.hasOwnProperty('BonusServerPassword')) mConfigJson.BonusServerPassword = objParams.BonusServerPassword;
            if (objParams.hasOwnProperty('UpperDomain')) mConfigJson.UpperDomain = objParams.UpperDomain;
            if (objParams.hasOwnProperty('SubDomains')) mConfigJson.SubDomains = objParams.SubDomains;
            if (objParams.hasOwnProperty('APITransferPoints')) mConfigJson.APITransferPoints = objParams.APITransferPoints;
            if (objParams.hasOwnProperty('CollectSubScale')) mConfigJson.CollectSubScale = objParams.CollectSubScale;
            if (objParams.hasOwnProperty('RabbitMQServer')) mConfigJson.RabbitMQServer = objParams.RabbitMQServer;
            if (objParams.hasOwnProperty('RabbitMQUserName')) mConfigJson.RabbitMQUserName = objParams.RabbitMQUserName;
            if (objParams.hasOwnProperty('RabbitMQPassword')) mConfigJson.RabbitMQPassword = objParams.RabbitMQPassword;
            if (objParams.hasOwnProperty('ConditionWinA')) mConfigJson.ConditionWinA = objParams.ConditionWinA;
            if (objParams.hasOwnProperty('ConditionWinB')) mConfigJson.ConditionWinB = objParams.ConditionWinB;
            if (objParams.hasOwnProperty('ConditionWinCR')) mConfigJson.ConditionWinCR = objParams.ConditionWinCR;

            const content = JSON.stringify(mConfigJson, null, 4);
            fs.writeFileSync(mConfigFileName, content);
        }
        catch(err) {
            Log.storeMsg(`Error writing config.bin: ${err.message}.`);
        }
    }
    
    static databases()
    {
        return mConfigJson["databases"];
    }

    static DBHost()
    {
        try
        {
            const dbs = this.databases();
            if (!dbs ||
                dbs.length == 0)
            {
                throw new Error("null databases");
            }

            const obj = dbs[0];
            if (obj == null) throw new Error("null obj in databases[0]");
            const domain = obj.domain;
            if (domain == null) throw new Error("null domain");
            return domain;
        }
        catch(err)
        {
            Log.storeMsg(err.message ? err.message : err);
            return "127.0.0.1";
        }
    }

    static DBPort()
    {
        try
        {
            const dbs = this.databases();
            if (!dbs ||
                dbs.length == 0)
            {
                throw new Error("null databases");
            }

            const obj = dbs[0];
            if (obj == null) throw new Error("null obj in databases[0]");
            const port = obj.port;
            if (port == null) throw new Error("null port");
            return parseInt(port);
        }
        catch(err)
        {
            Log.storeMsg(err.message ? err.message : err);
            return 27017;
        }
    }

    static WebSite()
    {
        return mConfigJson["WebSite"];
    }

    static RabbitMQServer()
    {
        return mConfigJson["RabbitMQServer"];
    }
    
    static RabbitMQUserName()
    {
        return mConfigJson["RabbitMQUserName"];
    }

    static RabbitMQPassword()
    {
        return mConfigJson["RabbitMQPassword"];
    }
    
    static BonusServerDomain()
    {
        return mConfigJson["BonusServerDomain"]
    }
    
    static BonusServerPort()
    {
        return mConfigJson["BonusServerPort"];
    }
    
    static BonusServerPassword()
    {
        return mConfigJson["BonusServerPassword"];
    }
    
    static UpperDomain()
    {
        return mConfigJson["UpperDomain"];
    }
    
    static SubDomains()
    {
        const list = [];
        const subDomains = mConfigJson["SubDomains"];
        if (subDomains &&
            subDomains.length > 0)
        {
            for (let i = 0; i < subDomains.length; i++)
            {
                const obj = subDomains[i];
                if (!obj) continue;
                const domain = obj["domain"];
                const password = obj["password"];
                const info = {
                    Domain: domain ? domain : "",
                    Password: password ? password : ""
                };
                list.push(info);
            }
        }
        return list;
    }

    static AddSubDomain(domain)
    {

    }
    
    static APITransferPoints()
    {
        return mConfigJson["TransferScores"];
    }
    
    static CollectSubScale()
    {
        return mConfigJson["CollectSubScale"];
    }
    
    static BetWinRule()
    {
        return mConfigJson["BetWinRule"];
    }
    
    static ConditionWinA()
    {
        return mConfigJson["ConditionWinA"];
    }
    
    static ConditionWinB()
    {
        return mConfigJson["ConditionWinB"];
    }
    
    static ConditionWinCR()
    {
        return mConfigJson["ConditionWinCR"];
    }
}

File: BonusServer/ConfigSetting.cs
using BonusServer.Services.RuleTrigger;
using FunLobbyUtils;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace BonusServer
{
    public class ConfigSetting : FunLobbyUtils.Config
    {
        public enum BETWIN_RULE
        {
            Rule_0,
            Rule_1
        }

        public class BonusLogin
        {
            public string Domain { get; set; }
            public string Password { get; set; }
        }

        public ConfigSetting(string configFileName, bool bEncodeContent = true) : base(configFileName, bEncodeContent)
        {
            this.storeConfig();
        }

        public void Reload()
        {
            mConfigJson = getJson();
        }

        public Dictionary<string, string> getContent()
        {
            string content = JsonConvert.SerializeObject(mConfigJson);
            JObject? obj = JsonConvert.DeserializeObject<JObject>(content);
            Dictionary<string, string> dicSettings = new Dictionary<string, string>();
            if (obj != null)
            {
                foreach (var item in obj)
                {
                    dicSettings[item.Key] = item.Value != null ? item.Value.ToString() : "";
                }
            }
            return dicSettings;
        }

        override protected JObject getJson()
        {
            int errCode = 0;
            JObject? configJson = null;
            try
            {
                configJson = base.getJson();

                // assign default value if it doesn't exist
                if (configJson.ContainsKey("databases") == false) configJson["databases"] = new JArray();
                JArray? databases = configJson["databases"] as JArray;
                if (databases == null) databases = new JArray();
                if (databases.Count == 0)
                {
                    JObject defaultDB = new JObject();
                    defaultDB["agent"] = "WinLobby";
                    defaultDB["type"] = "mongodb";
                    defaultDB["domain"] = "localhost";
                    defaultDB["port"] = 27017;

                    databases.Add(defaultDB);
                    configJson["databases"] = databases;
                }

                errCode = 100;

                if (configJson.ContainsKey("WebSite") == false) configJson["WebSite"] = "彩金";
                if (configJson.ContainsKey("BonusServerDomain") == false) configJson["BonusServerDomain"] = "http://localhost";
                if (configJson.ContainsKey("BonusServerPort") == false) configJson["BonusServerPort"] = 80;
                if (configJson.ContainsKey("BonusServerPassword") == false) configJson["BonusServerPassword"] = "localhost";
                if (configJson.ContainsKey("UpperDomain") == false) configJson["UpperDomain"] = "";
                if (configJson.ContainsKey("SubDomains") == false) configJson["SubDomains"] = new JArray();
                JArray? subDomains = configJson["SubDomains"] as JArray;
                if (subDomains == null) subDomains = new JArray();
                if (subDomains.Count == 0)
                {
                    JObject obj = new JObject();
                    obj["domain"] = "localhost";
                    obj["password"] = "localhost";
                    subDomains.Add(obj);
                    configJson["SubDomains"] = subDomains;
                }
                if (configJson.ContainsKey("TransferScores") == false) configJson["TransferScores"] = "http://slot888.tw/APIs/TransferPoints.ashx";
                if (configJson.ContainsKey("CollectSubScale") == false) configJson["CollectSubScale"] = 1.0f;
                if (configJson.ContainsKey("RabbitMQServer") == false) configJson["RabbitMQServer"] = "127.0.0.1";
                if (configJson.ContainsKey("RabbitMQUserName") == false) configJson["RabbitMQUserName"] = "guest";
                if (configJson.ContainsKey("RabbitMQPassword") == false) configJson["RabbitMQPassword"] = "guest";
                if (configJson.ContainsKey("BetWinRule") == false) configJson["BetWinRule"] = (int)BETWIN_RULE.Rule_0;

                errCode = 200;

                BETWIN_RULE rule = (BETWIN_RULE)(configJson["BetWinRule"].Value<int>());
                BonusRule? bonusRule = null;
                switch (rule)
                {
                    case BETWIN_RULE.Rule_0: bonusRule = new BonusRule_0(); break;
                    case BETWIN_RULE.Rule_1: bonusRule = new BonusRule_1(); break;
                    default: throw new Exception(string.Format("unknown rule {0}", rule.ToString()));
                }

                errCode = 300;
                try
                {
                    string? strContent = configJson["ConditionWinA"]?.Value<string>();
                    if (strContent == null) throw new Exception("null ConditionWinA");
                    bonusRule.ParseSettings(BonusRule.WIN_TYPE.WinA, strContent);
                }
                catch (Exception ex) { }

                errCode = 400;
                try
                {
                    string? strContent = configJson["ConditionWinB"]?.Value<string>();
                    if (strContent == null) throw new Exception("null ConditionWinB");
                    bonusRule.ParseSettings(BonusRule.WIN_TYPE.WinB, strContent);
                }
                catch (Exception ex) { }

                errCode = 500;
                try
                {
                    string? strContent = configJson["ConditionWinCR"]?.Value<string>();
                    if (strContent == null) throw new Exception("null ConditionWinCR");
                    bonusRule.ParseSettings(BonusRule.WIN_TYPE.WinCR, strContent);
                }
                catch (Exception ex) { }

                errCode = 600;
                // assign conditions to rule
                configJson["ConditionWinA"] = JsonConvert.SerializeObject(bonusRule.Condition_A.Settings());
                configJson["ConditionWinB"] = JsonConvert.SerializeObject(bonusRule.Condition_B.Settings());
                configJson["ConditionWinCR"] = JsonConvert.SerializeObject(bonusRule.Condition_CR.Settings());
            }
            catch (Exception ex)
            {
                string errMsg = string.Format("getJson got code:{0}, msg:{1}", errCode, ex.Message);
                throw new Exception(errMsg);
            }

            return configJson;
        }

        public string DBHost
        {
            get
            {
                try
                {
                    JArray? databases = mConfigJson["databases"] as JArray;
                    if (databases == null ||
                        databases.Count == 0)
                    {
                        throw new Exception("null databases");
                    }

                    JObject? obj = databases[0].Value<JObject>();
                    if (obj == null) throw new Exception("null obj in databases[0]");
                    string? domain = obj["domain"]?.Value<string>();
                    if (domain == null) throw new Exception("null domain");
                    return domain;
                }
                catch(Exception ex)
                {
                    Log.StoreMsg(ex.Message);
                    return "127.0.0.1";
                }
            }
        }
        public int DBPort
        {
            get
            {
                try
                {
                    JArray? databases = mConfigJson["databases"] as JArray;
                    if (databases == null ||
                        databases.Count == 0)
                    {
                        throw new Exception("null databases");
                    }

                    JObject? obj = databases[0].Value<JObject>();
                    if (obj == null) throw new Exception("null obj in databases[0]");
                    string? port = obj["port"]?.Value<string>();
                    if (port == null) throw new Exception("null port");
                    return int.Parse(port);
                }
                catch(Exception ex)
                {
                    Log.StoreMsg(ex.Message);
                    return 27017;
                }
            }
        }
        public JArray? DataBases { get { return mConfigJson["databases"]?.Value<JArray>(); } }

        public string? WebSite
        {
            get { return mConfigJson["WebSite"]?.Value<string>(); }
        }

        public string? RabbitMQServer
        {
            get { return mConfigJson["RabbitMQServer"]?.Value<string>(); }
            set { mConfigJson["RabbitMQServer"] = value; }
        }
        public string? RabbitMQUserName
        {
            get { return mConfigJson["RabbitMQUserName"]?.Value<string>(); }
            set { mConfigJson["RabbitMQUserName"] = value; }
        }
        public string? RabbitMQPassword
        {
            get { return mConfigJson["RabbitMQPassword"]?.Value<string>(); }
            set { mConfigJson["RabbitMQPassword"] = value; }
        }
        public string? BonusServerDomain
        {
            get { return mConfigJson["BonusServerDomain"]?.Value<string>(); }
            set { mConfigJson["BonusServerDomain"] = value; }
        }
        public int BonusServerPort
        {
            get { return mConfigJson["BonusServerPort"].Value<int>(); }
            set { mConfigJson["BonusServerPort"] = value; }
        }
        public string? BonusServerPassword
        {
            get { return mConfigJson["BonusServerPassword"]?.Value<string>(); }
            set { mConfigJson["BonusServerPassword"] = value; }
        }
        public string? UpperDomain
        {
            get { return mConfigJson["UpperDomain"]?.Value<string>(); }
            set { mConfigJson["UpperDomain"] = value; }
        }
        public List<BonusLogin> SubDomains
        {
            get
            {
                List<BonusLogin> list = new List<BonusLogin>();
                JArray? subDomains = mConfigJson["SubDomains"]?.Value<JArray>();
                if (subDomains != null &&
                    subDomains.Count > 0)
                {
                    for (int i = 0; i < subDomains.Count; i++)
                    {
                        JObject? obj = subDomains[i].Value<JObject>();
                        if (obj == null) continue;
                        string? domain = obj["domain"]?.Value<string>();
                        string? password = obj["password"]?.Value<string>();

                        BonusLogin info = new BonusLogin();
                        info.Domain = domain != null ? domain : "";
                        info.Password = password != null ? password : "";
                        list.Add(info);
                    }
                }
                return list;
            }
            set
            {
                JArray subDomains = new JArray();
                for (int i = 0; i < value.Count; i++)
                {
                    BonusLogin info = value[i];
                    JObject obj = new JObject();
                    obj["domain"] = info.Domain;
                    obj["password"] = info.Password;
                    subDomains.Add(obj);
                }
                mConfigJson["SubDomains"] = subDomains;
            }
        }
        public string? APITransferPoints
        {
            get { return mConfigJson["TransferScores"]?.Value<string>(); }
            set { mConfigJson["TransferScores"] = value; }
        }
        public float CollectSubScale
        {
            get { return mConfigJson["CollectSubScale"].Value<float>(); }
            set { mConfigJson["CollectSubScale"] = value; }
        }
        public BETWIN_RULE BetWinRule
        {
            get { return (BETWIN_RULE)mConfigJson["BetWinRule"].Value<int>(); }
            set { mConfigJson["BetWinRule"] = (int)value; }
        }
        public string? ConditionWinA
        {
            get { return mConfigJson["ConditionWinA"]?.Value<string>(); }
            set { mConfigJson["ConditionWinA"] = value; }
        }
        public string? ConditionWinB
        {
            get { return mConfigJson["ConditionWinB"]?.Value<string>(); }
            set { mConfigJson["ConditionWinB"] = value; }
        }
        public string? ConditionWinCR
        {
            get { return mConfigJson["ConditionWinCR"]?.Value<string>(); }
            set { mConfigJson["ConditionWinCR"] = value; }
        }
    }
}
